### PR TITLE
Add ZIP content view for issue #8051

### LINKED ISSUE #8051: Add a content view for zip files
#### Problem Description
It would be nice to easily see what files are included in a zip file download.

#### Proposal
Add a new content view for zip files (detectable via `content-type: application/zip`) that would give a list of files and their metadata.

#### Alternatives
Download the response and view it locally with something that reads zip files.

### PR DESCRIPTION
**Pull request by spider-yamet**
#### Description

- Implement ZipContentview with ZIP file parsing
- Display file metadata (names, sizes, compression methods, dates, comments)
- Add comprehensive test suite with 12 test cases covering all scenarios
- Register view in contentviews module
- Achieve 100% test coverage

Fixes #8051

#### Checklist

 - [x] I have updated tests where applicable.
 - [ ]  I have added an entry to the CHANGELOG.

### CODE REVIEW

**Comment by spider-yamet** [comment]
When mitmproxy intercepts a response with content-type application/zip, it will now display the ZIP archive contents in a readable format.

The view shows file names, compressed and uncompressed sizes, compression methods (like deflate or stored), modification dates when available, and file comments. It also displays overall archive statistics including total file count, total sizes, and compression ratio.

I tested this locally and all 15 tests pass. The implementation achieves 100% code coverage. I tested with various ZIP files including empty archives, nested directories, different compression methods, files with metadata like dates and comments, and error cases like corrupted files.

image

**Comment by normanr** [comment]
I think it would be nicer to format the output with one file per line, similar to how `unzip -l` does it, eg:
```
  Length      Date    Time    Name
---------  ---------- -----   ----
       38  2026-01-26 01:27   readme.txt
       33  2026-01-26 01:27   data/config.json
       32  2026-01-26 01:27   scripts/install.sh
       22  2024-01-15 14:30   important.txt
---------                     -------
      125                     4 files
```
I'm not sure if it should maybe use verbose output like `unzip -l -v`, which also shows the method, compressed size, compression ratio and crc (maybe there should be two different options for `ZIP` and `ZIP (verbose)`?)

**Comment by spider-yamet** [comment]
@normanr I updated the functionality and tested on my local end. It works correctly. Let me share the test result.

Standard View Output

```
  Length      Date    Time    Name
---------  ---------- -----   ----
       50  2026-01-27 02:38   readme.txt
       39  2026-01-27 02:38   src/main.py
       34  2026-01-27 02:38   data/config.json
       30  2026-01-27 02:38   LICENSE
---------                     -------
      153                      4 files
```


 Verbose View Output

The verbose view adds compression details, matching `unzip -l -v`:

```
 Length   Method    Size  Cmpr    Date    Time   CRC-32   Name
--------  ------  ------- ---- ---------- ----- --------  ----
      50  deflate      48   4% 2026-01-27 02:38 54c481a6   readme.txt
      39  deflate      41   0% 2026-01-27 02:38 70e89dd6   src/main.py
      34  deflate      33   3% 2026-01-27 02:38 9d00b829   data/config.json
      30  deflate      32   0% 2026-01-27 02:38 dc103c32   LICENSE
--------          ------- ---                            -------
     153              154   0%                             4 files
```

File: mitmproxy/contentviews/_view_zip.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+                    crc_str = f"{crc:08x}"
+                    row = (
+                        f"{length:>8}  {method:<{max_method}} {compressed:>7} {ratio_str:>4} "
+                        f"{date_pad:<10} {time_pad:<5} {crc_str}   {filename}"
+                    )
+                    lines.append(row)
+
+                # Summary separator
+                lines.append(
+                    "--------          ------- ---                            -------"
```
Comment (normanr) [inline mitmproxy/contentviews/_view_zip.py]: This got mangled. I think it's missing a single hyphen for the compression ratio column

File: mitmproxy/contentviews/_view_zip.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+                    time,
+                    filename,
+                ) in file_rows:
+                    date_pad = date if date else "          "
+                    time_pad = time if time else "     "
+                    ratio_str = f"{ratio:.0f}%" if ratio > 0 else "  0%"
+                    crc_str = f"{crc:08x}"
+                    row = (
+                        f"{length:>8}  {method:<{max_method}} {compressed:>7} {ratio_str:>4} "
+                        f"{date_pad:<10} {time_pad:<5} {crc_str}   {filename}"
```
Comment (normanr) [inline mitmproxy/contentviews/_view_zip.py]: One too many spaces between crc and filename

File: mitmproxy/contentviews/_view_zip.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+                    max(len(str(row[1])) for row in file_rows) if file_rows else 1
+                )
+                max_compressed = max(max_compressed, 9)
+                max_method = max(len(row[3]) for row in file_rows) if file_rows else 1
+                max_method = max(max_method, len("Method"))
+
+                # Build output matching unzip -l -v format
+                lines = []
+
+                # Header
```
Comment (normanr) [inline mitmproxy/contentviews/_view_zip.py]: I'm not sure what mitmproxy's policy on external libraries is (maybe @Prinzhorn can comment on it?), but maybe it would be better to use an existing library to format the data tables.

[tabulate](https://pypi.org/project/tabulate/), [PrettyTable](https://pypi.org/project/prettytable/), [asciitable](https://pypi.org/project/asciitable/) and [Rich tables](https://pypi.org/project/rich-tables/) are just some options (PrettyTable seems to be a good option?). None of them seem to support a final footer/total line, but you could probably simulate it with adding a divider row, and then the totals. The output doesn't need to match unzip exactly, but close is good enough.

File: mitmproxy/contentviews/_view_zip.py (hunk 1 above)
Comment (spider-yamet) [reply mitmproxy/contentviews/_view_zip.py]: It now has ---- (4 hyphens) for the compression ratio column, matching the main separator

File: mitmproxy/contentviews/_view_zip.py (hunk 2 above)
Comment (spider-yamet) [reply mitmproxy/contentviews/_view_zip.py]: Fixed CRC spacing: changed from 3 spaces to 2 spaces between CRC and filename

File: mitmproxy/contentviews/_view_zip.py (hunk 3 above)
Comment (spider-yamet) [reply mitmproxy/contentviews/_view_zip.py]: Hi @normanr, thanks for your opinion, now I have small question, should we wait for external library confirmation from @Prinzhorn or should I implement using PrettyTable now?

Regards

**Comment by mhils** [comment]
> I'm not sure what mitmproxy's policy on external libraries is (maybe @Prinzhorn can comment on it?), but maybe it would be better to use an existing library to format the data tables.

Not going to happen. The costs of additional dependencies greatly outweighs the benefits.

Beyond that, we need to stop some overengineering here:

- Please reduce this to one single contentview.
- Output as YAML. Filename and _maybe_ size, that's it. mitmproxy is an network debugger, not an archiving tool.

Thanks!

**Comment by spider-yamet** [comment]
Hi @mhils, thanks for your feedback. Now I reduced one output simplified for YAML format. Hope you can check about it.
Like below:
- filename: src/main.py
  size: 14
- filename: src/utils/helper.py
  size: 18
- filename: tests/test_main.py
  size: 16

Regards

File: mitmproxy/contentviews/_view_zip.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    def prettify(self, data: bytes, metadata: Metadata) -> str:
+        try:
+            with zipfile.ZipFile(io.BytesIO(data), "r") as zip_file:
+                info_list = zip_file.infolist()
+
+                if not info_list:
+                    return yaml_dumps([])
+
+                files = self._extract_file_info(info_list)
```
Comment (normanr) [inline mitmproxy/contentviews/_view_zip.py]: I would just inline the contents of _extract_file_info here. I don't think there's a good reason to have it as a separate method.

File: mitmproxy/contentviews/_view_zip.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        except Exception as e:
+            raise ValueError(f"Error parsing ZIP file: {e}")
+
+    def _extract_file_info(
+        self, info_list: list[zipfile.ZipInfo]
+    ) -> list[dict[str, int | str]]:
+        files = []
+        for info in info_list:
+            file_info: dict[str, int | str] = {"filename": info.filename}
+            if info.file_size > 0:
```
Comment (normanr) [inline mitmproxy/contentviews/_view_zip.py]: Why the conditional here? I think it would be simpler to just be consistent about including the size. Maybe it would be simpler (as mhils suggests) to drop the size, and only include the list of files. That would make the output simpler too, because it would just be a list of strings (instead of a list of dict)

Comment (spider-yamet) [reply mitmproxy/contentviews/_view_zip.py]: @normanr  @mhils I simplified the output format - Changed to a compact one-line-per-file format that includes both filename and size consistently:

- filename: readme.txt, size: 12
- filename: src/main.py, size: 11
- filename: data/config.json, size: 9
- filename: empty.txt, size: 0

File: mitmproxy/contentviews/_view_zip.py (hunk 4 above)
Comment (spider-yamet) [reply mitmproxy/contentviews/_view_zip.py]: Inlined the _extract_file_info method - Removed the separate helper method and put the logic directly in prettify().

File: mitmproxy/contentviews/_view_zip.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+    def prettify(self, data: bytes, metadata: Metadata) -> str:
+        try:
+            with zipfile.ZipFile(io.BytesIO(data), "r") as zip_file:
+                info_list = zip_file.infolist()
+                lines = [
+                    f"- filename: {info.filename}, size: {info.file_size}"
+                    for info in info_list
+                ]
+                return "\n".join(lines) + "\n" if lines else ""
```
Comment (mhils) [inline mitmproxy/contentviews/_view_zip.py]: I see some residual template injection risk here. Please collect into a list and then use 
https://github.com/mitmproxy/mitmproxy/blob/a7621021191011de7c95771459688de3ecd67c10/mitmproxy/contentviews/_utils.py#L74

File: mitmproxy/contentviews/_view_zip.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        except zipfile.BadZipFile:
+            raise ValueError("Invalid or corrupted ZIP file")
+        except Exception as e:
+            raise ValueError(f"Error parsing ZIP file: {e}")
```
Comment (mhils) [inline mitmproxy/contentviews/_view_zip.py]: What does the output look like for invalid zip files if we just remove the try/catch here?

File: test/mitmproxy/contentviews/test__view_zip.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+def test_view_zip_empty():
+    """Test empty ZIP file."""
+    buffer = io.BytesIO()
+    with zipfile.ZipFile(buffer, "w"):
+        pass
+    zip_data = buffer.getvalue()
+
+    result = zip.prettify(zip_data, meta("application/zip"))
+    assert result == ""
```
Comment (mhils) [inline test/mitmproxy/contentviews/test__view_zip.py]: This should probably return something like "(empty zip file)"

File: test/mitmproxy/contentviews/test__view_zip.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        def __enter__(self):
+            return self
+
+        def __exit__(self, *args):
+            return False
+
+        def infolist(self):
+            raise RuntimeError("Mocked infolist error")
+
+    monkeypatch.setattr(_view_zip.zipfile, "ZipFile", MockZipFile)
```
Comment (mhils) [inline test/mitmproxy/contentviews/test__view_zip.py]: Please reduce the testsuite so that we have full line coverage, but not more unless there are good reasons for it. Do one golden test that includes funny filenames (can we put a newline in it?) and then scratch the rest.

**Review by mhils — COMMENTED** [review]
Thanks, this looks _so_ much better!

File: mitmproxy/contentviews/_view_zip.py (hunk 5 above)
Comment (normanr) [reply mitmproxy/contentviews/_view_zip.py]: I think that's worse. What about something very simple like just: `{info.file_size:12} {info.filename}`? That's super simple and should work for most zip files without adding any complexity.

Comment (mhils) [reply mitmproxy/contentviews/_view_zip.py]: Just remove the size info, easy. :)

Comment (normanr) [reply mitmproxy/contentviews/_view_zip.py]: That works too. My biggest request is one line per-file, and a nice simple output. A list of just filenames achieves that.

Comment (spider-yamet) [reply mitmproxy/contentviews/_view_zip.py]: Simplified to just output a list of filenames

> - readme.txt
> - src/main.py
> - data/config.json
>

File: mitmproxy/contentviews/_view_zip.py (hunk 6 above)
Comment (spider-yamet) [reply mitmproxy/contentviews/_view_zip.py]: Now collecting filenames into a list and using yaml_dumps(): It will works now.

File: mitmproxy/contentviews/_view_zip.py (hunk 7 above)
Comment (spider-yamet) [reply mitmproxy/contentviews/_view_zip.py]: @mhils I removed this try/catch logic. The framework already handles exceptions in prettify_message()

File: test/mitmproxy/contentviews/test__view_zip.py (hunk 8 above)
Comment (spider-yamet) [reply test/mitmproxy/contentviews/test__view_zip.py]: Now returns "(empty zip file)" for empty archives.

File: test/mitmproxy/contentviews/test__view_zip.py (hunk 9 above)
Comment (spider-yamet) [reply test/mitmproxy/contentviews/test__view_zip.py]: Reduced to 3 minimal tests:
- test_view_zip() - Golden test with edge cases (spaces, nested dirs, newlines in filename, unicode)
- test_view_zip_empty() - Empty zip handling
- test_render_priority() - Priority logic

**Comment by spider-yamet** [comment]
@mhils @normanr I optimized code, please check my latest update.

Regards

**Comment by normanr** [comment]
looks fine to me

**Comment by spider-yamet** [comment]
Could you please confirm my PR if everything is fine? @normanr @mhils 

Regards

**Review by mhils — APPROVED** [review]
Look at this! Down from 200+ lines for the contenview to less than 20. Nice job! 🥳

**Comment by normanr** [comment]
Pity about the extensive commit message though :-/ that could have been a 1 liner, or a couple of lines at a push.
